Added specMACS cloud geometry markdown example and pictures (#99)
eurec4a/how_to_eurec4a: how_to_eurec4a/specmacs_cloud_geometry.md

ds = cat.HALO.specMACS.cloud_geometry[segment['segment_id']].to_dask()
ds
```

## Simple plots

We will start to do some first simple scatter plots of the cloud top heights (CTH). There are two CTH variables in the dataset: A normal 'height' variable and a variable called 'height_above_geoid'. The first is defined with respect to the WGS84 ellipsoid and the second with respect to the EGM2008 geoid. Although there is not a large difference between both in the EUREC4A area, we will exemplarily plot both.

```{code-cell} ipython3
%matplotlib inline
import matplotlib.pyplot as plt
import matplotlib
from mpl_toolkits.axes_grid1 import make_axes_locatable
```

```{code-cell} ipython3
#select only every 100th data point for a quick overview 
#(Attention: depending on the clouds there might be a lot of points in the dataset or not 
#such that this step might be useful or not)
ds_sel = ds.isel(sample=np.arange(0, ds.sizes['sample'], 100))
```

```{code-cell} ipython3
#load the data variables
lat = ds_sel['lat'].values
lon = ds_sel['lon'].values
height = ds_sel['height'].values
height_above_geoid = ds_sel['height_above_geoid'].values
```

```{code-cell} ipython3

> #create norm for scatter plot
> cmap = 'OrRd'
> vmin = np.min([np.min(height), np.min(height_above_geoid)])
> vmax = np.max([np.max(height), np.max(height_above_geoid)])
> norm = matplotlib.colors.Normalize(vmin=vmin, vmax=vmax)
> sm = matplotlib.cm.ScalarMappable(norm=norm, cmap=cmap)

Reviewer: I'd suggest to simplify the `min` and `max` calculation.
I'd also probably use `vmin` and `vmax` instead of `norm`. While this is less flexible, it's probably more commonly known. In order to get the colorbars right, the result of one of the `scatter` calls can be assigned to `sm`.
```suggestion
cmap = 'OrRd'
vmin = np.min([height, height_above_geoid])
vmax = np.max([height, height_above_geoid])
```

```

```{code-cell} ipython3
fig, axs = plt.subplots(nrows=2, ncols=1, figsize=(20, 8), sharex=True)

> axs[0].scatter(lon, lat, c=height, marker='o', s=2, edgecolor=None, linewidth=0, norm=norm, cmap=cmap)
> axs[0].set_title('cloud top height above WGS84 in m')
> axs[0].set_ylabel('latitude [°N]')
> divider = make_axes_locatable(axs[0])
> ax = divider.append_axes('right', size='40%', pad=0.05)
> ax.hist(height, bins=np.linspace(vmin, vmax, 100), density=True)
> ax.set_title('probability density')
> ax.yaxis.tick_right()
> ax.axvline(np.mean(height), linestyle='dashed', color='k', label='mean {:.2f} m'.format(np.mean(height)))
> ax.legend()
> axs[1].scatter(lon, lat, c=height_above_geoid, marker='o', s=2, edgecolor=None, linewidth=0, norm=norm, cmap=cmap)
> axs[1].set_title('cloud top height above sea surface in m')
> axs[1].set_ylabel('latitude [°N]')
> axs[1].set_xlabel('longitude [°E]')
> divider = make_axes_locatable(axs[1])
> ax = divider.append_axes('right', size='40%', pad=0.05)
> ax.hist(height_above_geoid, bins=np.linspace(vmin, vmax, 100), density=True)
> ax.set_title('probability density')
> ax.yaxis.tick_right()
> ax.axvline(np.mean(height_above_geoid), linestyle='dashed', color='k', label='mean {:.2f} m'.format(np.mean(height_above_geoid)))
> ax.legend()
> plt.colorbar(sm, ax=axs)

Reviewer: Much of this code is repeated for both rows of the plot. I suggest to put it inside a loop, to clarify how this code is executed twice.

For the `hist`, I also recommend to pass in the number of bins and the (equidistant) range explicitly. The argument is similar to `isel` above: this way, the `hist` function knows that the bins are equidistant and can use a more optimal code path.
```suggestion
for ax, var, ref in zip(axs, [height, height_above_geoid], ["WGS84", "sea surface"]):
    sm = ax.scatter(lon, lat, c=var, marker='o', s=2, edgecolor=None, linewidth=0, vmin=vmin, vmax=vmax, cmap=cmap)
    ax.set_title(f'cloud top height above {ref} in m')
    ax.set_ylabel('latitude [°N]')
    divider = make_axes_locatable(ax)
    ax2 = divider.append_axes('right', size='40%', pad=0.05)
    ax2.hist(height, bins=100, range=(vmin, vmax), density=True)
    ax2.set_title('probability density')
    ax2.yaxis.tick_right()
    ax2.axvline(np.mean(var), linestyle='dashed', color='k', label='mean {:.2f} m'.format(np.mean(var)))
    ax2.legend()

axs[1].set_xlabel('longitude [°E]')
plt.colorbar(sm, ax=axs)
```

> ```

Reviewer: I'd recommend to add a `None` statement (or a semicolon `;` after the last command) to suppres the printout of the colorbar object.
````suggestion
None
```
````

## Make a 3D plot

As the stereo points are points on the cloud surface located in 3D space, we can also do a 3D plot. We will use a few different viewing settings, which can also be individually adjusted using the `elev` and `azim` arguments of the `ax.view_init` setting.

```{code-cell} ipython3
import matplotlib.ticker as ticker
```

> ```{code-cell} ipython3
> #settings for plots 

Reviewer: Again: text is better placed in the markdown section.
```suggestion
settings for plots
```{code-cell} ipython3
```

> s = 0.1 #markersize

Reviewer: I suggest to use a telling variable name instead of a hard to remember variable name and a comment.
```suggestion
markersize = 0.1
```

cmap = 'viridis' #colormap
vmin = np.min(height)
vmax = np.max(height)

> norm = matplotlib.colors.Normalize(vmin=vmin, vmax=vmax)
> sm = matplotlib.cm.ScalarMappable(norm=norm, cmap=cmap)

Reviewer: (as before)
```suggestion
```

```

```{code-cell} ipython3
fig, axs = plt.subplots(2, 2, figsize=(14,12), subplot_kw=dict(projection='3d'))

> 

Reviewer: I was wondering if there could be a good way to reduce repetitions in in the code of the following plot, but I didn't come up with something much better. I'm however wondering if it might be more readable to do the actuall `scatter` in a loop, just to emphasize that it's really always the same data:

(this change would also account for `vmin`, `vmax` and `markersize`).

```suggestion

for ax in axs.ravel():
    sm = ax.scatter(lon, lat, height, c=height, marker='o', vmin=vmin, vmax=vmax, cmap=cmap, s=markersize, linewidths=None, edgecolors=None)

```

ax = axs[0,0]

> ax.scatter(lon, lat, height, c=height, marker='o', norm=norm, cmap=cmap, s=s, linewidths=None, edgecolors=None)

Reviewer: ```suggestion
```

ax.set_xlabel('longitude [°E]', labelpad=6.0)
ax.set_ylabel('latitude [°N]', labelpad=6.0)
ax.set_zlabel('height above WGS84 [m]', labelpad=6.0)
ax.view_init(elev=None, azim=None)
ax.set_title('3D stereo plot', y=1.02)

ax = axs[0,1]

> ax.scatter(lon, lat, height, c=height, marker='o', norm=norm, cmap=cmap, s=s, linewidths=None, edgecolors=None)

Reviewer: ```suggestion
```

ax.set_ylabel('latitude [°N]', labelpad=6.0)
ax.set_zlabel('height above WGS84 [m]', labelpad=6.0)
ax.view_init(elev=0, azim=0)
ax.set_title('latitude-height plane', y=1.02)
ax.xaxis.set_major_locator(ticker.NullLocator())

ax = axs[1,0]

> ax.scatter(lon, lat, height, c=height, marker='o', norm=norm, cmap=cmap, s=s, linewidths=None, edgecolors=None)

Reviewer: ```suggestion
```

ax.set_xlabel('longitude [°E]', labelpad=4.0)
ax.set_ylabel('latitude [°N]', labelpad=6.0)
ax.view_init(elev=90, azim=-90)
ax.set_title('longitude-latitude plane', y=1.02)
ax.zaxis.set_major_locator(ticker.NullLocator())

ax = axs[1,1]

> ax.scatter(lon, lat, height, c=height, marker='o', norm=norm, cmap=cmap, s=s, linewidths=None, edgecolors=None)

Reviewer: ```suggestion
```

ax.set_xlabel('longitude [°E]', labelpad=6.0)
ax.set_zlabel('height above WGS84 [m]', labelpad=10.0)
ax.view_init(elev=0, azim=-90)
ax.set_title('longitude-height plane', y=1.02)
ax.yaxis.set_major_locator(ticker.NullLocator())

plt.subplots_adjust(bottom=0.1, right=0.9)
cax = plt.axes([0.95, 0.15, 0.05, 0.7])

> fig.colorbar(sm, cax=cax, label='height above WGS84 [m]', fraction=0.02, shrink=2.0, aspect=40)

Reviewer: ```suggestion
fig.colorbar(sm, cax=cax, label='height above WGS84 [m]', fraction=0.02, shrink=2.0, aspect=40)
None
```

```

The 3D plots show the scene from different perspectives. It can be seen that the tracking algorithm finds points in both cloud layers down to a base height of approximately 1000m. Below, there are also some points which are probably outliers due to some wrong identifications.

+++

## Project points to specMACS measurements

> We will start to project the points found from the stereographic reconstruction to the specMACS maps which can be found [here](https://macsserver.physik.uni-muenchen.de/campaigns/EUREC4A/maps/). To do so, we need to import some further packages and we will define two functions to get the maps from the server.
> 
> ```{code-cell} ipython3
> from cartopy.io.img_tiles import GoogleWTS
> from PIL import Image
> import six 
> import cartopy.crs as ccrs
> ```
> 
> ```{code-cell} ipython3
> class DefaultTileWTS(GoogleWTS):
>     def get_image(self, tile):
>         if six.PY3:
>             from urllib.request import urlopen, Request, HTTPError, URLError
>         else:
>             from urllib2 import urlopen, Request, HTTPError, URLError
> 
>         url = self._image_url(tile)
>         try:
>             request = Request(url)
>             fh = urlopen(request)
>             im_data = six.BytesIO(fh.read())
>             fh.close()
>             img = Image.open(im_data)
> 
>         except (HTTPError, URLError) as err:
>             img = self._default_tile(tile)
> 
>         img = img.convert(self.desired_tile_form)
>         return img, self.tileextent(tile), 'lower'
> 
> class specMACSPTiles(DefaultTileWTS):
>     def __init__(self, flight_segment_id):
>         super(specMACSPTiles, self).__init__(desired_tile_form="RGBA")
>         self.flight_segment_id = flight_segment_id
>         self.empty_tile = Image.fromarray(np.full((256, 256, 4), (255, 255, 255, 0), dtype=np.uint8))
> 
>     def _default_tile(self, tile):
>         return self.empty_tile
> 
>     def _image_url(self, tile):
>         x, y, z = tile
>         return f"https://macsserver.physik.uni-muenchen.de/campaigns/EUREC4A/maps/tiles/{self.flight_segment_id}/{z}/{x}/{y}.png"
> ```
> 
> ```{code-cell} ipython3
> #get specMACS tile
> specmacs = specMACSPTiles(segment['segment_id'])
> ```

Reviewer: I'd try to keep the required code for the tiles rather minimal. I'd consider the tiles as a very nice background, but in principle unrelated to the purpose of this article, namely the stereo retrieval results. Thus I'd
* drop Python 2 support and the dependency on `six`
* merge the two classes into one
* hide the code by default
* move `cartopy.crs` import further down, as it's unrelated to this (now hidden) code block, but important for the following code

This change will also account for `segment_id` from the beginning.
````suggestion
We will start to project the points found from the stereographic reconstruction to pre-rendered specMACS [tile maps](https://en.wikipedia.org/wiki/Tiled_web_map) which can be found [here](https://macsserver.physik.uni-muenchen.de/campaigns/EUREC4A/maps/).
To make them available to `cartopy`, we need a little helper class `specMACSPTiles`, which mainly takes care of making unavailable tiles transparent (these are outside the viewing range of the instrument).

```{code-cell} ipython3
:tags: [hide-cell]

from cartopy.io.img_tiles import GoogleWTS
from PIL import Image

class specMACSPTiles(GoogleWTS):
    def __init__(self, flight_segment_id):
        super().__init__(desired_tile_form="RGBA")
        self.flight_segment_id = flight_segment_id
        self.empty_tile = Image.fromarray(np.full((256, 256, 4), (255, 255, 255, 0), dtype=np.uint8))

    def get_image(self, tile):
        from io import BytesIO
        from urllib.request import urlopen, Request, HTTPError, URLError

        url = self._image_url(tile)
        try:
            request = Request(url)
            fh = urlopen(request)
            im_data = BytesIO(fh.read())
            fh.close()
            img = Image.open(im_data)

        except (HTTPError, URLError) as err:
            img = self.empty_tile

        img = img.convert(self.desired_tile_form)
        return img, self.tileextent(tile), 'lower'

    def _image_url(self, tile):
        x, y, z = tile
        return f"https://macsserver.physik.uni-muenchen.de/campaigns/EUREC4A/maps/tiles/{self.flight_segment_id}/{z}/{x}/{y}.png"
```

We can now get the tile map for our selected flight segment:

```{code-cell} ipython3
specmacs = specMACSPTiles(segment_id)
```
````

> ```{code-cell} ipython3
> #get vmin and vmax using quantiles such that outliers will not influence the colorbar too much
> q = 0.98
> vmin = np.quantile(height, (1-q)/2)
> vmax = np.quantile(height, q+(1-q)/2)
> 
> cmap = 'OrRd'
> norm = matplotlib.colors.Normalize(vmin=vmin, vmax=vmax)
> sm = matplotlib.cm.ScalarMappable(norm=norm, cmap=cmap)
> ```

Reviewer: I'd suggest to move the comment into markdown, simplify the quantile computation and do the `vmin`, `vmax` trick again.
````suggestion
Get `vmin` and `vmax` using quantiles such that outliers will not influence the colorbar too much:
```{code-cell} ipython3
q = 0.01
vmin, vmax = np.quantile(height, [q, 1-q])
cmap = 'OrRd'
```
````

```{code-cell} ipython3

> fig = plt.figure(figsize=(16, 8), constrained_layout=True)
> ax = fig.add_subplot(1, 2, 1, projection=ccrs.PlateCarree())
> ax.set_extent([np.floor(np.min(lon)), np.ceil(np.max(lon)), np.floor(np.min(lat)), np.ceil(np.max(lat))], crs=ccrs.PlateCarree())
> ax.add_image(specmacs, 8) # the number is an integer and represents the detail level. It is logarithmic, small values are coarse, large values are fine.
> ax.set_axis_off()
> ax = fig.add_subplot(1, 2, 2, projection=ccrs.PlateCarree())
> ax.set_extent([np.floor(np.min(lon)), np.ceil(np.max(lon)), np.floor(np.min(lat)), np.ceil(np.max(lat))], crs=ccrs.PlateCarree())
> ax.add_image(specmacs, 8) # the number is an integer and represents the detail level. It is logarithmic, small values are coarse, large values are fine.
> ax.set_axis_off()
> im = ax.scatter(lon, lat, c=height, marker="o", s=0.4, edgecolor=None, linewidth=0, norm=norm, cmap=cmap)

Reviewer: Here, I'd again suggest to place the plotting inside a loop, just to have less repeated code.

Also, this is the place where `cartopy.crs` is needed, I'd place it here such that it's not hidden.

````suggestion
import cartopy.crs as ccrs

fig, axes = plt.subplots(1, 2, figsize=(16, 8), constrained_layout=True, subplot_kw={"projection": ccrs.PlateCarree()})
for ax in axes:
    ax.set_extent([np.floor(np.min(lon)), np.ceil(np.max(lon)), np.floor(np.min(lat)), np.ceil(np.max(lat))], crs=ccrs.PlateCarree())
    ax.add_image(specmacs, 8)  # the number is an integer and represents the detail level. It is logarithmic, small values are coarse, large values are fine.
    ax.set_axis_off()
im = ax.scatter(lon, lat, c=height, marker="o", s=0.4, edgecolor=None, linewidth=0, vmin=vmin, vmax=vmax, cmap=cmap)
````

> fig.colorbar(im, label='height above WGS84 [m]', extend='both', shrink=0.4)

Reviewer: ```suggestion
fig.colorbar(im, label='height above WGS84 [m]', extend='both', shrink=0.4)
None
```
